[ABLD-184] Add processing for --metadata files to gather licensing.

aiuto opened the pull request:
We were receiving them, but ignoring them.
- Add processing
- Convert 1 library (attr) to package_metadata rules to show the impact.
- Drive-by cleanup: Use csv writer

## smoke test
```
$ bazel build //deps:licenses
$ cat bazel-bin/deps/licenses.csv
core,@@+_repo_rules+bzip2//:license,@rules_license+//licenses/spdx:bzip2-1.0.6,"
-------------------------------------------------"
core,@@+_repo_rules+libffi//:license,@rules_license+//licenses/spdx:GPL-2.0,"libffi - Copyright (c) 1996-2024  Anthony Green, R"
core,@@+_repo_rules+nghttp2//:license,@rules_license+//licenses/spdx:MIT,"The MIT License

Copyright (c) 2012, 2014, 2015, 2"
core,@@+_repo_rules+xz//:license,@rules_license+//licenses/spdx:0BSD,"Permission to use, copy, modify, and/or distribute"
core,@@+_repo_rules+zlib//:license,@rules_license+//licenses/spdx:Zlib,"Copyright notice:

 (C) 1995-2022 Jean-loup Gailly"
core,pkg:https/download.savannah.nongnu.org/attr@2.5.1?url=https://download.savannah.nongnu.org/releases/attr/attr-%2.5.1.tar.xz,@rules_license+//licenses/spdx:LGPL-2.1,"Most components of the ""attr"" package are licensed"
```

Note the addition of attr in the list. The `origin` field is the PURL from `package_metadata`.

JSGette commented on compliance/gather_licenses.py:
```diff
@@ -38,32 +49,54 @@ def process_file(self, file, users):
+        if kind == "bazel-contrib.supply-chain.attribute.license":
+            self.process_license(attr)
+            return
+        print(f"Unknown attribute type: {kind}")
```
No error handling nor catching of faulty licenses?

aiuto replied:
We should fail hard. This is one of those impossible situations because it means the metadata is mal-formed.   I'll do that.

aiuto replied:
OMG.  That was 20 minutes of trying to rebase on top of main.

aiuto commented:
WTF???? 
NEED TO REBASE THIS. THERE ARE ONLY 3 FILES.  SORRY FOR THE NOISE

aiuto committed:
[ABLD-184] Add processing for --metadata files to gather licensing.

We were receiving them, but ignoring them.
- Add processing
- Convert 1 library (attr) to package_metadata rules to show the impact

```
$ bazel build //deps:licenses
$ cat bazel-bin/deps/licenses.csv
core,@@+_repo_rules+bzip2//:license,@rules_license+//licenses/spdx:bzip2-1.0.6,
-------------------------------------------------
core,@@+_repo_rules+libffi//:license,@rules_license+//licenses/spdx:GPL-2.0,libffi - Copyright (c) 1996-2024  Anthony Green, R
core,@@+_repo_rules+nghttp2//:license,@rules_license+//licenses/spdx:MIT,The MIT License

Copyright (c) 2012, 2014, 2015, 2
core,@@+_repo_rules+xz//:license,@rules_license+//licenses/spdx:0BSD,Permission to use, copy, modify, and/or distribute
core,@@+_repo_rules+zlib//:license,@rules_license+//licenses/spdx:Zlib,Copyright notice:

 (C) 1995-2022 Jean-loup Gailly
core,pkg:https/download.savannah.nongnu.org/attr@2.5.1?url=https://download.savannah.nongnu.org/releases/attr/attr-%2.5.1.tar.xz,@rules_license+//licenses/spdx:LGPL-2.1,Most components of the "attr" package are licensed
```

Note the addition of attr in the list. The `origin` field is the PURL from `package_metadata`.

merge upstream main

raise exception if the kind is unknown

aiuto commented:
/merge
